ably/event_emitter.go

```diff
@@ -0,0 +1,176 @@
+package ably
+
+import (
+	"runtime/debug"
+	"sync"
+)
+
+type eventEmitter struct {
+	sync.Mutex
+	listeners listenersForEvent
+	logger    Logger
+}
+
+type emitterEvent = interface{}
+type emitterData = interface{}
+
+type listenersForEvent map[emitterEvent]listenerSet
+type listenerSet map[*eventListener]struct{}
+
+type eventListener struct {
+	handler func(emitterData)
+	once    bool
+
+	queueMtx sync.Mutex
+	queue    []emitterData
+}
+
+func (l *eventListener) handle(e emitterData, log Logger) {
+	// The goroutine that finds the queue empty launches a goroutine for emptying it,
+	// ie. process its emitterData and afterwards any emitterData that concurrent
+	// goroutines may have left in the queue in the meantime.
+	//
+	// Other goroutines just enqueue their emitterData.
+	//
+	// This ensures our guarantee that a single handler runs concurrently with
+	// other handlers, but calls to the same handler are sequential and ordered,
+	// while emitting an event doesn't block the emitting goroutine.
+
+	var isBusy bool
+
+	l.queueMtx.Lock()
+	isBusy = len(l.queue) > 0
+	l.queue = append(l.queue, e)
+	l.queueMtx.Unlock()
+
+	if isBusy {
+		return
+	}
+
+	go func() {
+		done := false
+		for !done {
+			l.queueMtx.Lock()
+			e := l.queue[0]
+			l.queueMtx.Unlock()
+
+			safeHandle(e, l.handler, log)
+
+			l.queueMtx.Lock()
+			l.queue = l.queue[1:]
+			done = len(l.queue) == 0
+			l.queueMtx.Unlock()
+		}
+	}()
+}
+
+func safeHandle(e emitterData, handle func(emitterData), log Logger) {
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
+		}
+		log.Printf(LogError, "EventEmitter: panic in event handler: %v\n%s", r, debug.Stack())
+	}()
+
+	handle(e)
+}
+
+func newEventEmitter(logger Logger) *eventEmitter {
+	return &eventEmitter{
+		listeners: listenersForEvent{
+			nil: listenerSet{},
+		},
+		logger: logger,
+	}
+}
+
+// On registers an event listener. If event is nil, all events will trigger
+// the listener. Otherwise, it must be comparable to the eventEmitter's event
+// type, and only events equal to it will trigger the listener.
+//
+// It returns a function to deregister the listener.
+func (em *eventEmitter) On(event emitterEvent, handle func(emitterData)) (off func()) {
```

Comment:
> So shouldn't the type of `event` here be an enum of those values rather than an empty interface?

It is. This is the generic event emitter; the connection exposes a type-safe API for it with enums and whatnot.

> Whilst I'm aware other libraries use callbacks, I do think in the case of Go we should use channels instead.

I disagree. Channels are harder to get right, as they're a more primitive concurrency primitive than a callback-based event loop. If we just use channels, then users have to think about goroutine lifecycles, buffered or unbuffered, etc. It's just harder for little gain.

If you want to block, you can easily do something like your example on top of callbacks. We could even provide "wait" functions that wrap the callback-based event emitters to do this. But I don't see much use for it because these events are mainly "informative". You're not supposed to wait on a CONNECTED state for anything, really. You're supposed to just start subscribing and publishing and whatnot and things will get queued underneath. So it's just not that useful to block on these events. This is how the spec is designed.

Also, there's the issue of API uniformity. We want client libraries to be idiomatic, but also as close as possible to each other and to the abstract, OOP-ish API defined in the spec, which does specify callbacks here. We have the precedent of Ruby and Python, both using callbacks instead of fibers and generators, respectively.

There are of Go commonplace libraries that do callback-based (or delegate-based) APIs, such as `net/http.Server`, both in its core `Handler` and the `ConnState` hook. So I wouldn't just dismiss them as unidiomatic. It's definitely unidiomatic for IO, but not for events.

Anyway, this is my subjective opinion so I'm happy to do it the other way if you decide so.

Reply:
> So shouldn't the type of `event` here be an enum of those values rather than an empty interface?

It is. This is the generic event emitter; the connection exposes a type-safe API for it with enums and whatnot.

> Whilst I'm aware other libraries use callbacks, I do think in the case of Go we should use channels instead.

I disagree. Channels are harder to get right, as they're a more primitive concurrency primitive than a callback-based event loop. If we just use channels, then users have to think about goroutine lifecycles, buffered or unbuffered, etc. It's just harder for little gain.

If you want to block, you can easily do something like your example on top of callbacks. We could even provide "wait" functions that wrap the callback-based event emitters to do this. But I don't see much use for it because these events are mainly "informative". You're not supposed to wait on a CONNECTED state for anything, really. You're supposed to just start subscribing and publishing and whatnot and things will get queued underneath. So it's just not that useful to block on these events. This is how the spec is designed.

Also, there's the issue of API uniformity. We want client libraries to be idiomatic, but also as close as possible to each other and to the abstract, OOP-ish API defined in the spec, which does specify callbacks here. We have the precedent of Ruby and Python, both using callbacks instead of fibers and generators, respectively.

There are of Go commonplace libraries that do callback-based (or delegate-based) APIs, such as `net/http.Server`, both in its core `Handler` and the `ConnState` hook. So I wouldn't just dismiss them as unidiomatic. It's definitely unidiomatic for IO, but not for events.

Anyway, this is my subjective opinion so I'm happy to do it the other way if you decide so.

Reply:
My personal preference would be to make this as idiomatic as possible for the language this client library supports, thus aligning more with the suggestion from @lmars of using Go channels.

However, what has been made clear to me by both @mattheworiordan and @paddybyers is that the preference at Ably is to keep client libraries as close to the feature specification as possible. For that reason it is probably 'safer' to go with @tcard's existing choice of architecture. The lowest common denominator wins, as far as I see it, in order to achieve simplicity and (apparent?) parity across client library APIs.

There could be an opportunity to step back slightly and iterate on the feature specification where there may be gaps of recognition in respect of features available in 'newer' programming languages. I don't know that we have time to do this at the present time, but I fear that if we don't pretty soon then this will result in long term architectural stagnation of Ably's client library domain in general, in respect of our capability to respond to language and platform idomatic innovations.

I've [commented elsewhere](https://github.com/ably/docs/issues/791#issuecomment-572483717) about related concerns (as yet unresolved).

cc @SimonWoolf 🤓

Reply:
@tcard 

> This is the generic event emitter; the connection exposes a type-safe API for it with enums and whatnot.

But does it need to be _this_ generic (i.e. allowing a developer to pass in _anything_)? Same comment as https://github.com/ably/ably-go/pull/144#discussion_r404705037 I guess.

@tcard @QuintinWillison 

On the handler vs channel question:

My initial proposal for channels here was because I feel like ably-go was designed to be "low-level" and integrate nicely with Go's concurrency primitives (i.e. goroutines and channels), and I think this is evident by the library delivering both [messages](https://godoc.org/github.com/ably/ably-go/ably#Subscription.MessageChannel) and [presence](https://godoc.org/github.com/ably/ably-go/ably#Subscription.PresenceChannel) using Go channels.

I believe this design was chosen so that, rather than in languages like Ruby and JavaScript where Ably is expected to be running in an event loop and uses callbacks, in Go I can either build a handler-style event loop like:

```go
for {
	select {
	case msg := <-sub.MessageChannel():
		// handle a message
	case presence := <-sub.PresenceChannel():
		// handle a presence message
	case state := <-connState:
		// handle conn state
	case myEvent := <-myOtherChannelUnrelatedToAbly:
		// handle event unrelated to Ably
	}
}
```

Or, if say I just want to wait for some message on a channel before doing something else, well then I just receive on the channel, perhaps with my own timeout:

```go
sub, err := channel.Subscribe("some event i'm waiting for")
if err != nil {
	return err
}
defer sub.Close()
select {
case msg := <-sub.MessageChannel():
	// I got the message I wanted
case <-time.After(10 * time.Second):
	// I gave up waiting
}
```

Based on your comments though, it sounds like you are both saying we want to actively move away from this low-level, channel based design and move the API upwards a little to be more inline with other client libraries. If this is the case, then we should probably do it wholesale by providing a blocking Ably event loop that dispatches using callbacks and doesn't expose any channels.

What would this look like? Well, as you point out @tcard, this is pretty much what [`net/http.Server`](https://golang.org/pkg/net/http/#Server) does. Rather than creating a listener, running an accept loop and dispatching handlers, you just create a server, give it some handlers and (blockingly) start it. For Ably, this could look something like:

```go
ably := ably.NewRealtimeClient()

ably.Connection.On(ably.CONNECTED, func() {
	log.Printf("That was simple, you're now connected to Ably in realtime")
})

channel := ably.Channels.Get("quickstart")
channel.Subscribe("greeting", func(msg *ably.Message) {
	log.Printf("Received a greeting message in realtime: %v", msg.Data)
})

// the below will just block until ably.Stop() is called,
// or perhaps it could take a context
ably.Start()
```

I feel this brings the most parity with at least the Ruby and JavaScript implementations.

So, my suggestion here is to either:

1) adopt a callback design and use it consistently across the user facing API

2) stick with the current low-level design and use channels with the `EventEmitter`

@tcard a final point on this though:

> Channels are harder to get right, as they're a more primitive concurrency primitive than a callback-based event loop. If we just use channels, then users have to think about goroutine lifecycles, buffered or unbuffered, etc. It's just harder for little gain.

I have to disagree with this. So you're proposing that users implementing their own callback-based event loop is easier to get right than using channels? What does a user-implemented callback-based event loop even look like without using channels? What is being selected on?

Reply:
> I have to disagree with this. So you're proposing that users implementing their own callback-based event loop is easier to get right than using channels? What does a user-implemented callback-based event loop even look like without using channels? What is being selected on?

I do have to agree with @lmars that you're creating a lot of work for the user if you don't give them something they can select on, and all callbacks are from a goroutine that they do not control, which means they have to worry about synchronisation and locking in their own code.

Reply:
@QuintinWillison you pinged me on this, I assume because you want my opinion on the handler vs channel question?

If so... shrug. Language-idiomatic vs spec-compliant is a balancing act, each case is different. Subscribing to messages is one of those things that some languages may well have its own idiomatic way of doing things. I haven't used Go enough to have a strong opinion on this in Go, but in Elixir the most idiomatic thing would probably be to pass a process ID to the subscribe call, and messages would be sent to that process's inbox, and that'd be a perfectly reasonable API for an elixir client lib to use.

So, happy to go with whatever the experienced golang users think is best

(FWIW: the times when client lib developers tend to go off-spec in a way that comes back to bite us tend not to be things where they've carefully chosen a different API due to language differences. Rather they tend to be things where they've just added new methods/arguments/etc. that aren't in the spec because they think it'd be helpful, which then hamstring us when we add new features to the spec that conflict with them. E.g. ably-php adding a third `extras` object argument to publish(name, data) to make it more convenient to publish messages with extras, despite that not being in the spec, that then means that now we are officially adding a third argument to publish(name, data) for channelparams, we now have a backwards compatibility issue in ably-php in a library version that shouldn't have any breaking changes. That sort of thing. (Not to pick on ably-php, all the  client libs have been guilty of this at one time or another))

Reply:
Unless @lmars second proposal is a LOT of work, and worryingly non-Go-like, then I would prefer the second proposal which more closely matches ALL other SDKs.

Reply:
@lmars Sorry, I think I haven't given you enough context as a reviewer here.

This is part of [this task to have the library implement the spec](https://ably.atlassian.net/secure/RapidBoard.jspa?rapidView=22&projectKey=SDK&modal=detail&selectedIssue=FEA-391), whose first subtask is to have the library's API to match the spec's by wrapping the existing API, so that clients can then build atop without further breaking changes. This PR would be the first _sub_subtask of that subtask. (Yes, I've been asked to split that task already...)

So:

> 1. adopt a callback design and use it consistently across the user facing API

This is exactly the plan. Sorry I didn't make this clear.

> I have to disagree with this. So you're proposing that users implementing their own callback-based event loop is easier to get right than using channels? What does a user-implemented callback-based event loop even look like without using channels? What is being selected on?

I'm just saying channels provide more degrees of freedom than callbacks, so there's more ways to get them wrong. It's just a minor point, really, it's not like it's superhard to use channels for this. Perhaps a more compelling argument is that channels are a collaboration/synchronization primitive, and maybe using them here gives the wrong impression, as the library isn't trying to collaborate or synchronize with user code through event emitters; it's just informative, fire-and-forget from its point of view.

Anyway, this is all subjective and the stronger point I wanted to make is that I don't think channels provide enough advantages here to justify deviating from the callback-based API from the spec.

Reply:
> whose first subtask is to have the library's API to match the spec's by wrapping the existing API, so that clients can then build atop without further breaking changes. 

I realise you are not saying we need compatibility with existing users, but given 0.8 was "unstable", we have little traction on the realtime side, we really don't need to worry about backwards compatibility

Reply:
@SimonWoolf Those are good points, thanks. We've decided to deviate from the spec in a uniform manner in other places where it's worth it. Sadly, in this case, "whatever the experienced golang users think is best" isn't all that clear :)

Reply:
@tcard 

> This is part of this task to have the library implement the spec,

Ok that helps, thanks.

> > 1. adopt a callback design and use it consistently across the user facing API

> This is exactly the plan. Sorry I didn't make this clear.

Ok cool, so we're agreed then 🙂 

> first subtask is to have the library's API to match the spec's by wrapping the existing API, so that clients can then build atop without further breaking changes

Given the above decision means the API is going to be pretty different, I agree with @mattheworiordan on this one, we should just get on with shipping the new API and not worry about the existing one.

> the stronger point I wanted to make is that I don't think channels provide enough advantages here to justify deviating from the callback-based API from the spec.

Makes sense 👍

Reply:
I know I am an outsider here, but I wanted to link to this Gotime.fm podcast episode where they [discuss async design](https://overcast.fm/+GnKabFv3U/38:45), specifically what [Roberto Clapis](https://twitter.com/empijei) had to say:

> Every time I see a library that for example takes a channel and is going to send results over that channel… Or even worse, you call a function and you get a channel back - I’m not a big fan of that. Because I always have to read the code, for example, if I cancel the context, if they take a context… And are they going to check for cancelation at every send, or do I need to drain the channel? Will they block if the channel fills up? All these kinds of questions… I prefer my libraries stay synchronous. I don’t like promises.

Reply:
On the channel/handler question, I think go channels are nice when you're subscribing to 1 or 2 ably channels but there's no syntax for selecting over a slice of channels so this starts to get painful if you have many subscriptions or if you want to dynamically subscribe to different ably channels but still fit them into your event loop. You can do all those things straightforwardly using handler methods so I'm in favour of handlers.

I think Subscribe should then be blocking and we should be clear whether the handler methods are called synchronously or asynchronously.

Reply:
Thanks for your input @audiolion and @cruickshankpg! It looks like we're all on the same page here.